autoprecompiles/src/legacy_expression.rs

```diff
@@ -0,0 +1,225 @@
+    pub name: String,
+    /// Identifier for a polynomial reference, already contains
+    /// the element offset in case of an array element.
+    pub poly_id: PolyID,
+    pub next: bool,
+}
+
+impl std::fmt::Display for AlgebraicReference {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}{}", self.name, if self.next { "'" } else { "" },)
+    }
+}
+
+impl PartialOrd for AlgebraicReference {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for AlgebraicReference {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        (&self.poly_id, &self.next).cmp(&(&other.poly_id, &other.next))
+    }
+}
+
+impl PartialEq for AlgebraicReference {
+    fn eq(&self, other: &Self) -> bool {
+        self.poly_id == other.poly_id && self.next == other.next
+    }
+}
+
+impl Hash for AlgebraicReference {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.poly_id.hash(state);
+        self.next.hash(state);
+    }
+}
+
+/// Conversion to / from powdr_ast::analyzed::AlgebraicExpression.
+pub mod ast_compatibility {
+
+    use super::*;
+
+    impl From<AlgebraicReference> for powdr_ast::analyzed::AlgebraicReference {
+        fn from(reference: AlgebraicReference) -> Self {
+            powdr_ast::analyzed::AlgebraicReference {
+                name: reference.name,
+                poly_id: powdr_ast::analyzed::PolyID {
+                    id: reference.poly_id.id,
+                    ptype: match reference.poly_id.ptype {
+                        PolynomialType::Committed => powdr_ast::analyzed::PolynomialType::Committed,
+                        PolynomialType::Constant => powdr_ast::analyzed::PolynomialType::Constant,
+                        PolynomialType::Intermediate => {
+                            powdr_ast::analyzed::PolynomialType::Intermediate
+                        }
+                    },
+                },
+                next: reference.next,
+            }
+        }
+    }
+
+    impl From<powdr_ast::analyzed::AlgebraicReference> for AlgebraicReference {
+        fn from(reference: powdr_ast::analyzed::AlgebraicReference) -> Self {
+            AlgebraicReference {
+                name: reference.name,
+                poly_id: PolyID {
+                    id: reference.poly_id.id,
+                    ptype: match reference.poly_id.ptype {
+                        powdr_ast::analyzed::PolynomialType::Committed => PolynomialType::Committed,
+                        powdr_ast::analyzed::PolynomialType::Constant => PolynomialType::Constant,
+                        powdr_ast::analyzed::PolynomialType::Intermediate => {
+                            PolynomialType::Intermediate
+                        }
+                    },
+                },
+                next: reference.next,
+            }
+        }
+    }
+
+    pub trait CompatibleWithAstExpression<T>: Sized {
```

Comment:
Is this not `Into<powdr_ast::analyzed::AlgebraicExpression<T>> + TryFrom<powdr_ast::analyzed::AlgebraicExpression<T>>`?